uat/custom-components/client-java-paho/src/main/java/com/aws/greengrass/testing/mqtt5/client/paho/MqttConnectionImpl.java

```diff
@@ -507,4 +454,136 @@ private static List<Mqtt5Properties> getAckUserProperties(List<UserProperty> use
         }
         return userProperties;
     }
+
+    class MqttCallbackImpl implements MqttCallback {
+
+        @Override
+        @SuppressWarnings("PMD.AvoidCatchingGenericException")
+        public void disconnected(MqttDisconnectResponse mqttDisconnectResponse) {
+            GRPCClient.DisconnectInfo disconnectInfo = convertDisconnectPacket(mqttDisconnectResponse);
+
+            final String errorString = mqttDisconnectResponse.getException() == null
+                    ? null : mqttDisconnectResponse.getException().getMessage();
+
+            // only unsolicited disconnect
+            if (isClosing.get()) {
+                logger.atWarn().log("DISCONNECT event ignored due to shutdown initiated");
+            } else {
+                executorService.submit(() -> {
+                    try {
+                        grpcClient.onMqttDisconnect(connectionId, disconnectInfo, errorString);
+                    } catch (Exception ex) {
+                        logger.atError().withThrowable(ex).log("onMqttDisconnect failed");
+                    }
+                });
+            }
+        }
+
+        @Override
+        public void mqttErrorOccurred(org.eclipse.paho.mqttv5.common.MqttException e) {
+            logger.error("Client error with reason code {} and message '{}'", e.getReasonCode(), e.getMessage());
+        }
+
+        @Override
+        @SuppressWarnings("PMD.AvoidCatchingGenericException")
+        public void messageArrived(String topic, MqttMessage mqttMessage) throws Exception {
+            MqttConnectionImpl.this.messageArrived(topic, mqttMessage);
+        }
+
+        @Override
+        public void deliveryComplete(IMqttToken token) {
+            logger.atInfo().log("Delivery completion is {}", token.isComplete());
+        }
+
+        @Override
+        public void connectComplete(boolean reconnect, String s) {
+            logger.atInfo().log("Connection completed");
+        }
+
+        @Override
+        public void authPacketArrived(int i, MqttProperties mqttProperties) {
+            logger.atInfo().log("Connection completed");
+        }
+
+        private GRPCClient.DisconnectInfo convertDisconnectPacket(MqttDisconnectResponse response) {
```

Comment:
Probably should not on MqttCallbackImpl.
For received message you move all bussiness logic to class MqttConnectionImpl but for disconnect event you handle it inside  MqttCallbackImpl.
It looks strange.

Reply:
fixed in dfd2cb402305dcd47a03070e302a2a1f0592d8f7